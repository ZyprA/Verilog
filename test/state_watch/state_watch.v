`define NORMAL 2'b00
`define SEC 2'b01
`define MIN 2'b10
`define HOUR 2'b11

module state_watch(clk, reset, SW1, SW2, SW3, cs, sec_reset, min_inc, hour_inc);
    input clk, reset, SW1, SW2, SW3;
    output reg [1:0] cs;
    output sec_reset, min_inc, hour_inc;

    always @(posedge clk or negedge reset)
        if (!reset) cs <= `NORMAL;
        else
            case (cs)
                `NORMAL: 
                if (SW2)
                    cs <= `SEC;
                else
                    cs <= `NORMAL;
                `SEC:
                if (SW2)
                    cs <= `NORMAL;
                else if (SW3)
                    cs <= `HOUR;
                else
                    cs <= `SEC;
                `HOUR:
                if (SW2)
                    cs <= `NORMAL;
                else if (SW3)
                    cs <= `MIN;
                else
                    cs <= `HOUR;
                `MIN:
                if (SW2)
                    cs <= `NORMAL;
                else if (SW3)
                    cs <= `SEC;
                else
                    cs <= `MIN;
                default:
                    cs <= 2'bxx;
            endcase
    assign sec_reset = (cs == `SEC) && SW1;
    assign min_inc = (cs == `MIN) && SW1;
    assign hour_inc = (cs == `HOUR) && SW1;
endmodule

module state_watch_sim;
    reg clk, reset, SW1, SW2, SW3;
    wire [1:0] cs;
    wire sec_reset, min_inc, hour_inc;

    state_watch state_watch_inst(clk, reset, SW1, SW2, SW3, cs, sec_reset, min_inc, hour_inc);

    initial begin
        $dumpfile("state_watch.vcd");
        $dumpvars(0, state_watch_sim);
    end

    initial begin
        clk = 0;
        forever #11 clk = ~clk;
    end

    initial begin
        reset = 0;
        SW1 = 0;
        SW2 = 0;
        SW3 = 0;
        #100 reset = 1;
        #100 SW2 = 1;
        #30 SW2 = 0;
        #100 SW3 = 1;
        #30 SW3 = 0;
        #100 SW3 = 1;
        #30 SW3 = 0;
        #100 SW3 = 1; 
        #30 SW3 = 0;
        #100 SW1 = 1;
        #100 $finish;
    end

endmodule