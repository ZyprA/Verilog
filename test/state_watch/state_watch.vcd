$date
	Sun Feb  2 20:50:47 2025
$end
$version
	Icarus Verilog
$end
$timescale
	1s
$end
$scope module state_watch_sim $end
$var wire 1 ! sec_reset $end
$var wire 1 " min_inc $end
$var wire 1 # hour_inc $end
$var wire 2 $ cs [1:0] $end
$var reg 1 % SW1 $end
$var reg 1 & SW2 $end
$var reg 1 ' SW3 $end
$var reg 1 ( clk $end
$var reg 1 ) reset $end
$scope module state_watch_inst $end
$var wire 1 % SW1 $end
$var wire 1 & SW2 $end
$var wire 1 ' SW3 $end
$var wire 1 ( clk $end
$var wire 1 # hour_inc $end
$var wire 1 " min_inc $end
$var wire 1 ) reset $end
$var wire 1 ! sec_reset $end
$var reg 2 * cs [1:0] $end
$upscope $end
$upscope $end
$enddefinitions $end
$comment Show the parameter values. $end
$dumpall
$end
#0
$dumpvars
b0 *
0)
0(
0'
0&
0%
b0 $
0#
0"
0!
$end
#11
1(
#22
0(
#33
1(
#44
0(
#55
1(
#66
0(
#77
1(
#88
0(
#99
1(
#100
1)
#110
0(
#121
1(
#132
0(
#143
1(
#154
0(
#165
1(
#176
0(
#187
1(
#198
0(
#200
1&
#209
b1 $
b1 *
1(
#220
0(
#230
0&
#231
1(
#242
0(
#253
1(
#264
0(
#275
1(
#286
0(
#297
1(
#308
0(
#319
1(
#330
0(
1'
#341
b11 $
b11 *
1(
#352
0(
#360
0'
#363
1(
#374
0(
#385
1(
#396
0(
#407
1(
#418
0(
#429
1(
#440
0(
#451
1(
#460
1'
#462
0(
#473
b10 $
b10 *
1(
#484
0(
#490
0'
#495
1(
#506
0(
#517
1(
#528
0(
#539
1(
#550
0(
#561
1(
#572
0(
#583
1(
#590
1'
#594
0(
#605
b1 $
b1 *
1(
#616
0(
#620
0'
#627
1(
#638
0(
#649
1(
#660
0(
#671
1(
#682
0(
#693
1(
#704
0(
#715
1(
#720
1!
1%
#726
0(
#737
1(
#748
0(
#759
1(
#770
0(
#781
1(
#792
0(
#803
1(
#814
0(
#820
