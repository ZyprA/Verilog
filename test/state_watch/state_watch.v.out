#! /opt/homebrew/Cellar/icarus-verilog/12.0/bin/vvp
:ivl_version "12.0 (stable)";
:ivl_delay_selection "TYPICAL";
:vpi_time_precision + 0;
:vpi_module "/opt/homebrew/Cellar/icarus-verilog/12.0/lib/ivl/system.vpi";
:vpi_module "/opt/homebrew/Cellar/icarus-verilog/12.0/lib/ivl/vhdl_sys.vpi";
:vpi_module "/opt/homebrew/Cellar/icarus-verilog/12.0/lib/ivl/vhdl_textio.vpi";
:vpi_module "/opt/homebrew/Cellar/icarus-verilog/12.0/lib/ivl/v2005_math.vpi";
:vpi_module "/opt/homebrew/Cellar/icarus-verilog/12.0/lib/ivl/va_math.vpi";
S_0x13cf11350 .scope module, "state_watch_sim" "state_watch_sim" 2 49;
 .timescale 0 0;
v0x13cf24400_0 .var "SW1", 0 0;
v0x13cf24490_0 .var "SW2", 0 0;
v0x13cf24520_0 .var "SW3", 0 0;
v0x13cf245d0_0 .var "clk", 0 0;
v0x13cf24680_0 .net "cs", 1 0, v0x13cf23ff0_0;  1 drivers
v0x13cf24750_0 .net "hour_inc", 0 0, L_0x13cf24fd0;  1 drivers
v0x13cf24800_0 .net "min_inc", 0 0, L_0x13cf24dc0;  1 drivers
v0x13cf248b0_0 .var "reset", 0 0;
v0x13cf24960_0 .net "sec_reset", 0 0, L_0x13cf24bb0;  1 drivers
S_0x13cf10900 .scope module, "state_watch_inst" "state_watch" 2 54, 2 6 0, S_0x13cf11350;
 .timescale 0 0;
    .port_info 0 /INPUT 1 "clk";
    .port_info 1 /INPUT 1 "reset";
    .port_info 2 /INPUT 1 "SW1";
    .port_info 3 /INPUT 1 "SW2";
    .port_info 4 /INPUT 1 "SW3";
    .port_info 5 /OUTPUT 2 "cs";
    .port_info 6 /OUTPUT 1 "sec_reset";
    .port_info 7 /OUTPUT 1 "min_inc";
    .port_info 8 /OUTPUT 1 "hour_inc";
L_0x13cf24bb0 .functor AND 1, L_0x13cf24a90, v0x13cf24400_0, C4<1>, C4<1>;
L_0x13cf24dc0 .functor AND 1, L_0x13cf24ce0, v0x13cf24400_0, C4<1>, C4<1>;
L_0x13cf24fd0 .functor AND 1, L_0x13cf24eb0, v0x13cf24400_0, C4<1>, C4<1>;
v0x13cf14700_0 .net "SW1", 0 0, v0x13cf24400_0;  1 drivers
v0x13cf23990_0 .net "SW2", 0 0, v0x13cf24490_0;  1 drivers
v0x13cf23a30_0 .net "SW3", 0 0, v0x13cf24520_0;  1 drivers
L_0x140078010 .functor BUFT 1, C4<01>, C4<0>, C4<0>, C4<0>;
v0x13cf23ac0_0 .net/2u *"_ivl_0", 1 0, L_0x140078010;  1 drivers
L_0x1400780a0 .functor BUFT 1, C4<11>, C4<0>, C4<0>, C4<0>;
v0x13cf23b70_0 .net/2u *"_ivl_12", 1 0, L_0x1400780a0;  1 drivers
v0x13cf23c60_0 .net *"_ivl_14", 0 0, L_0x13cf24eb0;  1 drivers
v0x13cf23d00_0 .net *"_ivl_2", 0 0, L_0x13cf24a90;  1 drivers
L_0x140078058 .functor BUFT 1, C4<10>, C4<0>, C4<0>, C4<0>;
v0x13cf23da0_0 .net/2u *"_ivl_6", 1 0, L_0x140078058;  1 drivers
v0x13cf23e50_0 .net *"_ivl_8", 0 0, L_0x13cf24ce0;  1 drivers
v0x13cf23f60_0 .net "clk", 0 0, v0x13cf245d0_0;  1 drivers
v0x13cf23ff0_0 .var "cs", 1 0;
v0x13cf240a0_0 .net "hour_inc", 0 0, L_0x13cf24fd0;  alias, 1 drivers
v0x13cf24140_0 .net "min_inc", 0 0, L_0x13cf24dc0;  alias, 1 drivers
v0x13cf241e0_0 .net "reset", 0 0, v0x13cf248b0_0;  1 drivers
v0x13cf24280_0 .net "sec_reset", 0 0, L_0x13cf24bb0;  alias, 1 drivers
E_0x13cf0a100/0 .event negedge, v0x13cf241e0_0;
E_0x13cf0a100/1 .event posedge, v0x13cf23f60_0;
E_0x13cf0a100 .event/or E_0x13cf0a100/0, E_0x13cf0a100/1;
L_0x13cf24a90 .cmp/eq 2, v0x13cf23ff0_0, L_0x140078010;
L_0x13cf24ce0 .cmp/eq 2, v0x13cf23ff0_0, L_0x140078058;
L_0x13cf24eb0 .cmp/eq 2, v0x13cf23ff0_0, L_0x1400780a0;
    .scope S_0x13cf10900;
T_0 ;
    %wait E_0x13cf0a100;
    %load/vec4 v0x13cf241e0_0;
    %nor/r;
    %flag_set/vec4 8;
    %jmp/0xz  T_0.0, 8;
    %pushi/vec4 0, 0, 2;
    %assign/vec4 v0x13cf23ff0_0, 0;
    %jmp T_0.1;
T_0.0 ;
    %load/vec4 v0x13cf23ff0_0;
    %dup/vec4;
    %pushi/vec4 0, 0, 2;
    %cmp/u;
    %jmp/1 T_0.2, 6;
    %dup/vec4;
    %pushi/vec4 1, 0, 2;
    %cmp/u;
    %jmp/1 T_0.3, 6;
    %dup/vec4;
    %pushi/vec4 3, 0, 2;
    %cmp/u;
    %jmp/1 T_0.4, 6;
    %dup/vec4;
    %pushi/vec4 2, 0, 2;
    %cmp/u;
    %jmp/1 T_0.5, 6;
    %pushi/vec4 3, 3, 2;
    %assign/vec4 v0x13cf23ff0_0, 0;
    %jmp T_0.7;
T_0.2 ;
    %load/vec4 v0x13cf23990_0;
    %flag_set/vec4 8;
    %jmp/0xz  T_0.8, 8;
    %pushi/vec4 1, 0, 2;
    %assign/vec4 v0x13cf23ff0_0, 0;
    %jmp T_0.9;
T_0.8 ;
    %pushi/vec4 0, 0, 2;
    %assign/vec4 v0x13cf23ff0_0, 0;
T_0.9 ;
    %jmp T_0.7;
T_0.3 ;
    %load/vec4 v0x13cf23990_0;
    %flag_set/vec4 8;
    %jmp/0xz  T_0.10, 8;
    %pushi/vec4 0, 0, 2;
    %assign/vec4 v0x13cf23ff0_0, 0;
    %jmp T_0.11;
T_0.10 ;
    %load/vec4 v0x13cf23a30_0;
    %flag_set/vec4 8;
    %jmp/0xz  T_0.12, 8;
    %pushi/vec4 3, 0, 2;
    %assign/vec4 v0x13cf23ff0_0, 0;
    %jmp T_0.13;
T_0.12 ;
    %pushi/vec4 1, 0, 2;
    %assign/vec4 v0x13cf23ff0_0, 0;
T_0.13 ;
T_0.11 ;
    %jmp T_0.7;
T_0.4 ;
    %load/vec4 v0x13cf23990_0;
    %flag_set/vec4 8;
    %jmp/0xz  T_0.14, 8;
    %pushi/vec4 0, 0, 2;
    %assign/vec4 v0x13cf23ff0_0, 0;
    %jmp T_0.15;
T_0.14 ;
    %load/vec4 v0x13cf23a30_0;
    %flag_set/vec4 8;
    %jmp/0xz  T_0.16, 8;
    %pushi/vec4 2, 0, 2;
    %assign/vec4 v0x13cf23ff0_0, 0;
    %jmp T_0.17;
T_0.16 ;
    %pushi/vec4 3, 0, 2;
    %assign/vec4 v0x13cf23ff0_0, 0;
T_0.17 ;
T_0.15 ;
    %jmp T_0.7;
T_0.5 ;
    %load/vec4 v0x13cf23990_0;
    %flag_set/vec4 8;
    %jmp/0xz  T_0.18, 8;
    %pushi/vec4 0, 0, 2;
    %assign/vec4 v0x13cf23ff0_0, 0;
    %jmp T_0.19;
T_0.18 ;
    %load/vec4 v0x13cf23a30_0;
    %flag_set/vec4 8;
    %jmp/0xz  T_0.20, 8;
    %pushi/vec4 1, 0, 2;
    %assign/vec4 v0x13cf23ff0_0, 0;
    %jmp T_0.21;
T_0.20 ;
    %pushi/vec4 2, 0, 2;
    %assign/vec4 v0x13cf23ff0_0, 0;
T_0.21 ;
T_0.19 ;
    %jmp T_0.7;
T_0.7 ;
    %pop/vec4 1;
T_0.1 ;
    %jmp T_0;
    .thread T_0;
    .scope S_0x13cf11350;
T_1 ;
    %vpi_call 2 57 "$dumpfile", "state_watch.vcd" {0 0 0};
    %vpi_call 2 58 "$dumpvars", 32'sb00000000000000000000000000000000, S_0x13cf11350 {0 0 0};
    %end;
    .thread T_1;
    .scope S_0x13cf11350;
T_2 ;
    %pushi/vec4 0, 0, 1;
    %store/vec4 v0x13cf245d0_0, 0, 1;
T_2.0 ;
    %delay 11, 0;
    %load/vec4 v0x13cf245d0_0;
    %inv;
    %store/vec4 v0x13cf245d0_0, 0, 1;
    %jmp T_2.0;
    %end;
    .thread T_2;
    .scope S_0x13cf11350;
T_3 ;
    %pushi/vec4 0, 0, 1;
    %store/vec4 v0x13cf248b0_0, 0, 1;
    %pushi/vec4 0, 0, 1;
    %store/vec4 v0x13cf24400_0, 0, 1;
    %pushi/vec4 0, 0, 1;
    %store/vec4 v0x13cf24490_0, 0, 1;
    %pushi/vec4 0, 0, 1;
    %store/vec4 v0x13cf24520_0, 0, 1;
    %delay 100, 0;
    %pushi/vec4 1, 0, 1;
    %store/vec4 v0x13cf248b0_0, 0, 1;
    %delay 100, 0;
    %pushi/vec4 1, 0, 1;
    %store/vec4 v0x13cf24490_0, 0, 1;
    %delay 30, 0;
    %pushi/vec4 0, 0, 1;
    %store/vec4 v0x13cf24490_0, 0, 1;
    %delay 100, 0;
    %pushi/vec4 1, 0, 1;
    %store/vec4 v0x13cf24520_0, 0, 1;
    %delay 30, 0;
    %pushi/vec4 0, 0, 1;
    %store/vec4 v0x13cf24520_0, 0, 1;
    %delay 100, 0;
    %pushi/vec4 1, 0, 1;
    %store/vec4 v0x13cf24520_0, 0, 1;
    %delay 30, 0;
    %pushi/vec4 0, 0, 1;
    %store/vec4 v0x13cf24520_0, 0, 1;
    %delay 100, 0;
    %pushi/vec4 1, 0, 1;
    %store/vec4 v0x13cf24520_0, 0, 1;
    %delay 30, 0;
    %pushi/vec4 0, 0, 1;
    %store/vec4 v0x13cf24520_0, 0, 1;
    %delay 100, 0;
    %pushi/vec4 1, 0, 1;
    %store/vec4 v0x13cf24400_0, 0, 1;
    %delay 100, 0;
    %vpi_call 2 81 "$finish" {0 0 0};
    %end;
    .thread T_3;
# The file index is used to find the file name in the following table.
:file_names 3;
    "N/A";
    "<interactive>";
    "state_watch.v";
